corehq/apps/accounting/migrations/0023__simplify__credit_line__product_type.py

```diff
@@ -0,0 +1,117 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+from datetime import date
+
+from django.db import models, migrations
+from django.db.models import F, Q
+
+from corehq.sql_db.operations import HqRunPython
+
+
+def confirm_no_mismatched_subscription_credit(apps, schema_editor):
+    if apps.get_model('accounting', 'CreditLine').objects.filter(
+        product_type__isnull=False,
+        subscription__isnull=False,
+    ).exclude(
+        product_type=F('subscription__plan_version__product_rate__product__product_type'),
+    ).exclude(
+        balance=0,
+    ).exists():
+        raise Exception("""
+
+Running this migration will make previously inaccessible subscription credit usable.
+
+For production environments, reconcile credit lines on a case by case basis.
+
+For dev environments, simply delete problematic credit lines:
+
+from django.db.models import F
+from corehq.apps.accounting.models import CreditAdjustment, CreditLine
+problematic_credit_lines = CreditLine.objects.filter(
+    product_type__isnull=False,
+    subscription__isnull=False,
+).exclude(
+    product_type=F('subscription__plan_version__product_rate__product__product_type'),
+).exclude(
+    balance=0,
+)
+CreditAdjustment.objects.filter(credit_line__in=problematic_credit_lines).delete()
+CreditAdjustment.objects.filter(related_credit__in=problematic_credit_lines).delete()
+problematic_credit_lines.delete()
+            """
+        )
+
+
+def confirm_no_mismatched_account_credit_can_be_invoiced_automatically(apps, schema_editor):
+    previous_invoicing_date = date(date.today().year, date.today().month, 1)
+    subscriptions_in_future_invoicing_periods = apps.get_model('accounting', 'Subscription').objects.exclude(
+        date_end__lte=previous_invoicing_date,
+    )
+    if filter(
+        lambda sub: apps.get_model('accounting', 'CreditLine').objects.filter(
+            account=sub.account,
+            product_type__isnull=False,
+        ).exclude(
+            Q(product_type='') | Q(product_type=sub.plan_version.product_rate.product.product_type)
+        ).exclude(
+            balance=0
+        ).exists(),
+        subscriptions_in_future_invoicing_periods
+    ):
+        raise Exception("""
+
+Running this migration may allow subscriptions in future invoicing periods access to
+product credit not originally allocated to the subscription's product type.
+
+For production environments, reconcile credit lines on a case by case basis.
+
+For dev environments, simply delete problematic credit lines:
+
+from datetime import date
+from django.db.models import Q
+from corehq.apps.accounting.models import CreditAdjustment, CreditLine, Subscription
+previous_invoicing_date = date(date.today().year, date.today().month, 1)
+subscriptions_in_future_invoicing_periods = Subscription.objects.exclude(
```

Comment:
oops, should have ignored suppressed subscriptions

https://github.com/dimagi/commcare-hq/pull/11272